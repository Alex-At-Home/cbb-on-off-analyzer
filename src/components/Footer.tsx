// React imports:
import React, { useState } from 'react';

// Next imports:
import { NextPage } from 'next';

// Lodash:
import _ from "lodash";

// Bootstrap imports:
import 'bootstrap/dist/css/bootstrap.min.css';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';

// Data imports:
import { dataLastUpdated } from '../utils/internal-data/dataLastUpdated';

type Props = {
  readonly server: string,
  readonly year?: string,
  readonly gender?: string
}

const Footer: React.FunctionComponent<Props> = ({server, gender, year}) => {

  const emailAddress = "bWFpbHRvOmhvb3AuZXhwbG9yZXJAZ21haWwuY29t";
  const twitterAddress = "aHR0cHM6Ly90d2l0dGVyLmNvbS9JdHNBVGVycF9DQkI=";

  const lastUpdated = _.flow([
    (maybeYear: string, maybeGender: string) => [ maybeYear || "", maybeGender || "" ],
    (yearGender: string) => dataLastUpdated[`${yearGender[1]}_${yearGender[0]}`],
    (lastUpdate: string | undefined) => lastUpdate ?
      new Date(parseInt(lastUpdate)*1000).toString() : "unknown"
  ]);

  const onMouseOver = (encoded: string) => (event: any) => {
    if (!_.startsWith(event.target.href, "mailto") &&
        !_.startsWith(event.target.href, "https"))
    {
      event.target.href = atob(encoded);
    }
  }

  // (Display no footer on the non-public site:)
  if (!_.startsWith(server, "cbb-on-off-analyzer")) return <Container><Row>
      <Col>
        <i><small>This year's data last updated: [{lastUpdated(year, gender)}]</small></i>
      </Col>
      <Col>
        <span className="float-right">
          <i><small>
          PbP events from <a href="https://stats.ncaa.org" target="_new">stats.ncaa.org</a>
          </small></i>
        </span>
      </Col>
    </Row>
    <Row>
      <Col>
        <i><small>It's a beta, so let me know if you see anything weird:&nbsp;
        <a href={twitterAddress} target="_new"
          onMouseOver={onMouseOver(twitterAddress)}
        >
        twitter
        </a>&nbsp;/&nbsp;
        <a href={emailAddress} target="_new"
          onMouseOver={onMouseOver(emailAddress)}
        >
        email
        </a></small></i>
      </Col>
      <Col>
        <span className="float-right">
          <i><small>
          SoS stats from <a href="https://kenpom.com" target="_new">kenpom.com</a> and <a href="https://herhoopstats.com" target="_new">herhoopstats.com</a>
          </small></i>
        </span>
      </Col>
    </Row></Container>;
  else
    return <Container></Container>;
}

export default Footer;
