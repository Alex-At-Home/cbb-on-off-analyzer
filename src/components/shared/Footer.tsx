// React imports:
import React, { useState } from "react";

// Next imports:
import { NextPage } from "next";

// Lodash:
import _ from "lodash";

// Bootstrap imports:
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";

// Data imports:
import { dataLastUpdated } from "../../utils/internal-data/dataLastUpdated";
import { DateUtils } from "../../utils/DateUtils";
import { OverlayTrigger, Tooltip } from "react-bootstrap";
import { ClientRequestCache } from "../../utils/ClientRequestCache";

type Props = {
  readonly server: string;
  readonly year?: string;
  readonly gender?: string;
  readonly dateOverride?: number;
};

const Footer: React.FunctionComponent<Props> = ({
  server,
  gender,
  year,
  dateOverride,
}) => {
  const emailAddress = "bWFpbHRvOmhvb3AuZXhwbG9yZXJAZ21haWwuY29t";
  const twitterAddress = "aHR0cHM6Ly90d2l0dGVyLmNvbS9JdHNBVGVycF9DQkI=";

  const lastUpdated = _.flow([
    (maybeYear: string, maybeGender: string) => [
      maybeYear || "",
      maybeGender || "",
    ],
    (yearGender: string) =>
      dataLastUpdated[`${yearGender[1]}_${yearGender[0]}`],
    (lastUpdate: string | undefined) =>
      lastUpdate || dateOverride
        ? new Date(
            (dateOverride || parseInt(lastUpdate || "0")) * 1000
          ).toString()
        : "unknown",
    // Some browsers (cough firefox cough), show the timezone in full format which is annoyingly long
    // so turn the timezone into initials in that case
    (lastUpdate: string) =>
      lastUpdate.replace(
        /[(]([A-Z][a-z][^)]*)+[)]/,
        (match: string, ...args: any[]) => {
          return `(${args[0]
            .split(" ")
            .map((w: string) => w?.[0] || "")
            .join("")})`;
        }
      ),
  ]);

  const onMouseOver = (encoded: string) => (event: any) => {
    if (
      !_.startsWith(event.target.href, "mailto:") &&
      !_.startsWith(event.target.href, "https://twitter.com")
    ) {
      event.target.href = atob(encoded);
    }
  };

  const publicSite = !_.startsWith(server, "cbb-on-off-analyzer");

  const womenStatsSource =
    _.startsWith(year, "2018") || _.startsWith(year, "2019") ? (
      <a href="https://herhoopstats.com" target="_blank">
        herhoopstats.com
      </a>
    ) : (year || "2024") >= "2024" ? (
      <a href="https://barttorvik.com/ncaaw/trank.php" target="_blank">
        barttorvik.com
      </a>
    ) : (
      <a href="https://masseyratings.com/cbw/ncaa-d1/ratings" target="_blank">
        masseyratings.com
      </a>
    );

  // (only display twitter on public site, for analytics purposes)
  return (
    <Container>
      <Row>
        <Col>
          <i>
            <small>
              Data updated: [{lastUpdated(year, gender)}] [
              <OverlayTrigger
                placement="auto"
                overlay={
                  <Tooltip id="clearCache">
                    If the data appears out of date trying clicking here to
                    clear the cache and refresh the page
                  </Tooltip>
                }
              >
                <a
                  href="#"
                  onClick={(e) => {
                    ClientRequestCache.clearCache();
                    window.location.reload();
                  }}
                >
                  Clear Cache
                </a>
              </OverlayTrigger>
              ]
            </small>
          </i>
        </Col>
        <Col>
          <span className="float-right">
            <i>
              <small>
                PbP events and roster info from{" "}
                <a href="https://stats.ncaa.org" target="_blank">
                  stats.ncaa.org
                </a>{" "}
                and{" "}
                <a href="https://collegebasketballdata.com" target="_blank">
                  collegebasketballdata.com
                </a>
              </small>
            </i>
          </span>
        </Col>
      </Row>
      <Row>
        <Col>
          <i>
            <small>
              It's a beta(-ish!), so let me know if you see anything
              weird:&nbsp;
              {publicSite ? (
                <span>
                  <a
                    href={twitterAddress}
                    target="_blank"
                    onMouseOver={onMouseOver(twitterAddress)}
                  >
                    twitter
                  </a>
                  &nbsp;/&nbsp;
                </span>
              ) : null}
              <a
                href={emailAddress}
                target="_blank"
                onMouseOver={onMouseOver(emailAddress)}
              >
                email
              </a>
            </small>
          </i>
        </Col>
        <Col>
          <span className="float-right">
            <i>
              <small>
                SoS stats with kind permission from{" "}
                <a href="https://kenpom.com" target="_blank">
                  kenpom.com
                </a>{" "}
                and {womenStatsSource}
              </small>
            </i>
          </span>
        </Col>
      </Row>
    </Container>
  );
};

export default Footer;
