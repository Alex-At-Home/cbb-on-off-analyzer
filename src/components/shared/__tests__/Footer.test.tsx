import React from 'react';
import Footer from '../Footer';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';


describe("Footer", () => {
  test("Footer - should create snapshot", () => {
    const component = renderer.create(<Footer
      server="server"
      year="2019/20"
      gender="Men"
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
