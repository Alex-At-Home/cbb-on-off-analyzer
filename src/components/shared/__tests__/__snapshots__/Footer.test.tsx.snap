// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Footer Footer - should create snapshot 1`] = `
<div
  className="container"
>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <i>
        <small>
          Selected year's data last updated: [
          Wed Mar 11 2020 19:18:35 GMT-0400 (EDT)
          ]
        </small>
      </i>
    </div>
    <div
      className="col"
    >
      <span
        className="float-right"
      >
        <i>
          <small>
            PbP events from 
            <a
              href="https://stats.ncaa.org"
              target="_new"
            >
              stats.ncaa.org
            </a>
          </small>
        </i>
      </span>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col"
    >
      <i>
        <small>
          It's a beta, so let me know if you see anything weird: 
          <span>
            <a
              href="aHR0cHM6Ly90d2l0dGVyLmNvbS9JdHNBVGVycF9DQkI="
              onMouseOver={[Function]}
              target="_new"
            >
              twitter
            </a>
             / 
          </span>
          <a
            href="bWFpbHRvOmhvb3AuZXhwbG9yZXJAZ21haWwuY29t"
            onMouseOver={[Function]}
            target="_new"
          >
            email
          </a>
        </small>
      </i>
    </div>
    <div
      className="col"
    >
      <span
        className="float-right"
      >
        <i>
          <small>
            SoS stats with kind permission from 
            <a
              href="https://kenpom.com"
              target="_new"
            >
              kenpom.com
            </a>
             and 
            <a
              href="https://herhoopstats.com"
              target="_new"
            >
              herhoopstats.com
            </a>
          </small>
        </i>
      </span>
    </div>
  </div>
</div>
`;
